### PR TITLE
Add AWS guide

### PR DESCRIPTION
**Pull request by djosborne**
Add a new AWS guide that covers all the "gotchas" of running Calico in AWS

### CODE REVIEW

File: master/reference/public-cloud/aws.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+IP-in-IP is enabled on a per-pool basis with the following command:
+
+```yaml
+cat << EOF | calicoctl apply -f -
+apiVersion: v1
+kind: ipPool
+metadata:
+  cidr: 10.1.0.0/16
+spec:
+  ipip:
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: You actually need to enabled nat-outgoing for this as well if you want container->host traffic to work.

I think we should just be recommending to always enabled nat-outgoing when `ipip` is enabled.

File: master/reference/public-cloud/aws.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+---
+
+Calico is designed to provide high performance massively scalable virtual
+networking not only for private data centers, but also within a
+public cloud such as Amazon Web Services (AWS).  The following information
+details necessary modifications to your standard AWS deployment to allow
+Calico to function successfully.
+
+
+## Prerequisite: Host Connectivity
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: I think this would be better as:

```
## Prerequisites

### Host Connectivity
```

File: master/reference/public-cloud/aws.md [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+sends or receives.
+
+Since Calico allocates a unique IP to each container, container traffic leaving
+and entering the EC2 instance will have an "unrecognized" source address
+and traffic will be dropped.
+
+This behavior can be disabled for inter-VPC workloads,
+and can be circumvented for cross-VPC workloads.
+
+#### Inter-VPC: Disable Src/Dest Checks
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: I think you mean "intra-VPC".  I'd just call this "Within a single VPC"

File: master/reference/public-cloud/aws.md [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+See the [AWS guide on disabling src/dest checks for an EC2 instance](http://docs.aws.amazon.com/AmazonVPC/latest/UserGuide/VPC_NAT_Instance.html#EIP_Disable_SrcDestCheck).
+
+#### Cross-VPC: Enable IP-in-IP
+
+Unfortunately, cross-VPC src/dest checks cannot be disabled in AWS.
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: This isn't quite right.  The src/dst check is just a per-node RPF check.

Cross-subnet traffic actually goes through a router, which isn't aware of pod traffic, and so tunneling is required.

File: master/reference/public-cloud/aws.md [hunk 5]
```diff
@@ -0,0 +1,152 @@
+---
+title: AWS
+---
+
+Calico is designed to provide high performance massively scalable virtual
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: We should mention that Calico lets you bypass the 50-node limit that exists when using VPC routes (e.g Flannel AWS VPC mode)

File: master/reference/public-cloud/aws.md (hunk 2 above)
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: We should document the exact holes that need to be poked in the AWS security-group in order for Calico to work.

These will vary based on whether you're disabling src/dst check or if you're using `ipip`

File: master/reference/public-cloud/aws.md [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+By default, each EC2 instance performs source/destination checks on inbound and outbound traffic.
+This means that the instance must be the source or destination of any traffic it
+sends or receives.
+
+Since Calico allocates a unique IP to each container, container traffic leaving
+and entering the EC2 instance will have an "unrecognized" source address
+and traffic will be dropped.
+
+This behavior can be disabled for inter-VPC workloads,
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: s/inter/intra

Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: (Or just way "within a single VPC")

Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: Maybe just:

"If your traffic is all within a single VPC, you can simply disable this behavior.  For cross-VPC traffic, you will need to use encapsulation.  Calico supports both as discussed below."

File: master/reference/public-cloud/aws.md [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+  cidr: 10.1.0.0/16
+spec:
+  ipip:
+    enabled: true
+EOF
+```
+
+## Allowing Workload-to-WAN Traffic
+
+In a private cloud where operators maintain full control of the networking
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: This isn't _always_ the case.

File: master/reference/public-cloud/aws.md [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+apiVersion: v1
+kind: ipPool
+metadata:
+  cidr: 10.1.0.0/16
+spec:
+  nat-outgoing: true
+EOF
+```
+
+## (Optional) Modifying MTU for Performance Boost
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: s/Performance Boost/Better Performance

File: master/reference/public-cloud/aws.md [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+modify the MTU of the veth interfaces of your Calico containers.
+
+> **WARNING**: You must run these commands on ALL of the hosts and containers in your
+> deployment. If not all of your instances are jumbo frame instances or if you
+> do not modify MTU on all containers and hosts, you may  experience unexpected
+> behavior and/or packet loss.
+
+#### Modify Calico interface MTU on hosts
+
+First, modify the MTU on the `cali######`
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: Oh man these are really manual instructions.

Can't we say "use the CNI plugin's MTU parameter" to do this?

File: master/reference/public-cloud/aws.md (hunk 1 above)
Comment (djosborne) [inline master/reference/public-cloud/aws.md]: Why would intra-VPC container-to-host need nat-outgoing? As long as the target host has Calico routing and src/dest checks have been disabled the connection shouldn't be dropped

File: master/reference/public-cloud/aws.md (hunk 2 above)
Comment (djosborne) [inline master/reference/public-cloud/aws.md]: But it's the only prerequisite...

File: master/reference/public-cloud/aws.md (hunk 1 above)
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: You need nat-outgoing whenever `ipip` is enabled.  Otherwise traffic will leave with a src: podIP dst: hostIP and be dropped by the RPF.

ipip only handles pod->pod traffic.

File: master/reference/public-cloud/aws.md (hunk 5 above)
Comment (djosborne) [inline master/reference/public-cloud/aws.md]: [Flannel's AWS backend doc](https://github.com/coreos/flannel/blob/master/README.md#backends) states that it still limits you to 50:

>Note: Currently, AWS limits the number of entries per route table to 50.

Comment (djosborne) [inline master/reference/public-cloud/aws.md]: Ah. you mean when using `ipip`. Updated to point this out.

File: master/reference/public-cloud/aws.md [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+dropped as it passes through the ec2 hosts to the container.
+
+This reverse-path filtering is enabled by default, but can be disabled on each
+host, allowing calico traffic to pass through unaffected.
+**Ensure you follow
+[AWS documentation on disabling source/destination checks](http://docs.aws.amazon.com/AmazonVPC/latest/UserGuide/VPC_NAT_Instance.html#EIP_Disable_SrcDestCheck).**
+
+#### Allow Traffic Across Different Subnets
+
+Container traffic that traverses different subnets will past through
```
Comment (2blesd2bstresd) [inline master/reference/public-cloud/aws.md]: "pass"

File: master/reference/public-cloud/aws.md [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+
+1. Masquerade Container Traffic
+
+   Container traffic can be masqueraded with its host IP so that the VPC
+   never sees the container IPs, allowing standard ec2 routing to take over.
+
+   Turn on traffic masquerading in pool settings by enabling:
+
+   - `ipip` for container-to-container traffic.
+   - `nat-outgoing` for container-to-ec2-instancce traffic.
```
Comment (2blesd2bstresd) [inline master/reference/public-cloud/aws.md]: "instance"

**Comment by 2blesd2bstresd** [comment]
Worth mentioning that the kubeadm setup guide correctly configures ipip and nat-outgoing.
http://kubernetes.io/docs/getting-started-guides/kubeadm/

File: master/reference/public-cloud/aws.md [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+title: AWS
+---
+
+Calico is designed to provide high performance massively scalable virtual
+networking for containers in public cloud environments such as
+Amazon Web Services (AWS).
+
+## Why Use Calico in AWS
+
+#### Container Policy
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: Network Policy for Containers

File: master/reference/public-cloud/aws.md [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+
+While AWS Security Groups control network traffic sent to ec2 instances, it
+does not cover containers running on hosts. Calico implements fine-grain policy
+for containers which dynamically applies to containers as they dynamically
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: This sentence needs rewording - it says containers and dynamically too many times.

How about "Calico automatically implements fine-grained, dynamic network policy for containers."

File: master/reference/public-cloud/aws.md [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+move around the cluster.
+
+#### No Overlays
+
+Calico minimizes the need for overlays by assigning each container
+its own IP and allowing containers to talk to one another in a pure IP network.
+
+#### Break Through the '50  Route Limit'
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: I think this should be the first in the list, and should be titled "No 50 node limit" or "Massive Scale" or something.

File: master/reference/public-cloud/aws.md [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+
+[AWS VPC Routing tables are limited to 50 programmable routes](http://docs.aws.amazon.com/AmazonVPC/latest/UserGuide/VPC_Appendix_Limits.html#vpc-limits-route-tables).
+While this might limit other networking plugins, Calico supports
+ip-over-ip encapsulation to circumvent manually programming container routes in
+the VPC routing table altogether.
+
+## How to Run Calico in AWS
+
+Depending on the architecture of your AWS deployment, some of the following
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: I'd suggest rewording this section to be something more along the lines of "To deploy Calico in AWS, you must ensure that the proper security group rules have been made and that traffic between containers on different hosts is not dropped by the VPC.  There are a few different options for doing this depending on your deployment."

Then maybe we can split the sections below into "### Security Group Configuration" and "### Cross-host communication"

File: master/reference/public-cloud/aws.md [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+modifications may need to be made to a stock calico cluster in order for
+it to function properly.
+
+#### Allow BGP Peering in Security Groups
+
+To allow Calico containers to communicate across hosts on AWS, each host
+should have IP connectivity with one another and be able to BGP peer.
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: "be able to BGP peer" is weird.  We also want to make it clear that this is only required when you're using BGP, not when you're not using BGP.

How about "When using Calico BGP to distribute container routes, you need to ensure that your security group allows BGP protocol traffic."

File: master/reference/public-cloud/aws.md [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+
+To allow this BGP peering, ensure your security group settings for each
+calico host allows incoming tcp connections to port 179.
+
+You can check if your hosts have successfully established BGP sessions with
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: Ideally this becomes a link to a section in a doc which is "How do I tell Calico is working", but fine for now.  Let's just keep this spot in mind.

File: master/reference/public-cloud/aws.md [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+one another using `calicoctl node status`.
+
+#### Allow Traffic Within a Single VPC Subnet
+
+By default, each EC2 instance performs source/destination checks on inbound
+and outbound traffic. This ensures that the instance is the source or
+destination of any traffic it sends or receives.
+
+Since Calico allocates a unique IP to each container using pools AWS is not aware
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: rather than say "pools AWS is not aware of" I think we should say "Since Calico allocates a unique IP address to each container that isn't the IP of its host, and since Calico doesn't encapsulate container traffic by default, traffic to/from containers will be dropped by the AWS src/dst checks."

File: master/reference/public-cloud/aws.md [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+To allow Calico containers to communicate across hosts on AWS, each host
+should have IP connectivity with one another and be able to BGP peer.
+
+To allow this BGP peering, ensure your security group settings for each
+calico host allows incoming tcp connections to port 179.
+
+You can check if your hosts have successfully established BGP sessions with
+one another using `calicoctl node status`.
+
+#### Allow Traffic Within a Single VPC Subnet
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: It says it in the title, but otherwise this section doesn't make it clear that this will only work for traffic within a single VPC subnet and why.

File: master/reference/public-cloud/aws.md [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+Since Calico allocates a unique IP to each container using pools AWS is not aware
+of, container traffic will have an "unrecognized" source address and will be
+dropped as it passes through the ec2 hosts to the container.
+
+This reverse-path filtering is enabled by default, but can be disabled on each
+host, allowing calico traffic to pass through unaffected.
+**Ensure you follow
+[AWS documentation on disabling source/destination checks](http://docs.aws.amazon.com/AmazonVPC/latest/UserGuide/VPC_NAT_Instance.html#EIP_Disable_SrcDestCheck).**
+
+#### Allow Traffic Across Different Subnets
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: VPC Subnets

File: master/reference/public-cloud/aws.md (hunk 10 above)
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: VPC subnets

File: master/reference/public-cloud/aws.md [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+
+Container traffic that traverses different subnets will past through
+a VPC routing table. If this VPC routing table does not know how to route
+Calico IPs, it will drop traffic. There are two ways two prevent this from
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: I'm not a huge fan of the phrase "Calico IPs".  I'd rather we say "container/pod IPs" 

"it will drop traffic _to containers/pods_"

File: master/reference/public-cloud/aws.md [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+happening:
+
+1. Masquerade Container Traffic
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: This should say "Encapsulate Container Traffic".

File: master/reference/public-cloud/aws.md [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+   never sees the container IPs, allowing standard ec2 routing to take over.
+
+   Turn on traffic masquerading in pool settings by enabling:
+
+   - `ipip` for container-to-container traffic.
+   - `nat-outgoing` for container-to-ec2-instancce traffic.
+
+   See [pool configuration](todo) for information on how to enable this for pools.
+
+2. Configure the VPC Routing Table
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: You should note that this approach incurs a 50 node limit due to the size of the routing table, and as such is not recommended for large deployments.

File: master/reference/public-cloud/aws.md [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+
+   Container traffic can be masqueraded with its host IP so that the VPC
+   never sees the container IPs, allowing standard ec2 routing to take over.
+
+   Turn on traffic masquerading in pool settings by enabling:
+
+   - `ipip` for container-to-container traffic.
+   - `nat-outgoing` for container-to-ec2-instancce traffic.
+
+   See [pool configuration](todo) for information on how to enable this for pools.
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: We can at least link to the pool reference guide for now and update when we create a pool configuration guide.

File: master/reference/public-cloud/aws.md [hunk 25]
```diff
@@ -0,0 +1,115 @@
...
+   1. Deploy [canal][canal] and select the [aws-vpc backend][aws-vpc-backend],
+      which interacts with AWS api's to automatically program the routing table.
+
+      >Note: Doing so will limit you to just 50 routes.
+
+   2. Manually program the routing table (not recommended).
+
+#### Allow Traffic Across Different VPCs
+
+AWS drops non-masqueraded container traffic when it passes through multiple VPC's.
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: drop "non-masqueraded"

"when it passes through multiple VPC's" sort of implies it is transiting through a VPC. This should just say "when it passes between two VPCs"

Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: And I think it is VPCs rather than VPC's.

File: master/reference/public-cloud/aws.md [hunk 26]
```diff
@@ -0,0 +1,115 @@
...
+   ways.
+
+   1. Deploy [canal][canal] and select the [aws-vpc backend][aws-vpc-backend],
+      which interacts with AWS api's to automatically program the routing table.
+
+      >Note: Doing so will limit you to just 50 routes.
+
+   2. Manually program the routing table (not recommended).
+
+#### Allow Traffic Across Different VPCs
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: Maybe we should say "Peered VPCs" instead of different, to make it clear that this is traffic between two VPCs which have been peered? And then maybe in the body include a link to VPC peering?

File: master/reference/public-cloud/aws.md [hunk 27]
```diff
@@ -0,0 +1,115 @@
...
+To circumvent this issue, enable `ipip` and `nat-outgoing` on your pools.
+
+#### Allow Workload-to-WAN Traffic
+
+In a private cloud where operators maintain full control of the networking
+fabric, NAT'ing typically takes place at the edge router.
+
+AWS does not support NAT'ing outbound traffic from unknown sources (i.e. calico IPs).
+
+Fortunately, calico can perform NAT on outbound traffic on each compute node by
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: Uppercase "Calico", drop "Fortunately", "simply".

pool -> Calico IP pool

File: master/reference/public-cloud/aws.md [hunk 28]
```diff
@@ -0,0 +1,115 @@
...
+simply enabling `nat-outgoing` on each pool.
+
+## Next Steps
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: Not sure a "next steps" is appropriate for this doc - it's not really a step-by-step.

File: master/reference/public-cloud/aws.md [hunk 29]
```diff
@@ -0,0 +1,113 @@
...
+   See [pool configuration]({{site.baseurl}}/{{page.version}}/reference/calicoctl/resources/ippool) for information on how to enable this for IP pools.
+
+2. Configure the VPC Routing Table
+
+   Alternatively, the VPC routing table can be programmed with container/pod IP
+   routes so that it can route container traffic natively.
+   This can be accomplished in two ways:
+
+   1. Deploy [canal][canal] and select the [aws-vpc backend][aws-vpc-backend],
+      which interacts with AWS api's to automatically program the routing table.
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: api's -> APIs (no apostrophe)

File: master/reference/public-cloud/aws.md [hunk 30]
```diff
@@ -0,0 +1,113 @@
...
+the VPC routing table altogether.
+
+## How to Run Calico in AWS
+
+To deploy Calico in AWS, you must ensure that the proper security group rules
+have been made and that traffic between containers on different hosts is not
+dropped by the VPC. There are a few different options for doing this depending
+on your deployment.
+
+#### Configure Security Groups for BGP
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: It's not just BGP that needs security group allowances.  You should also mention that when using `ipip` you need to allow IP protocol 4 in the security group or else `ipip` traffic will be dropped.

File: master/reference/public-cloud/aws.md [hunk 31]
```diff
@@ -0,0 +1,113 @@
...
+
+   See [pool configuration]({{site.baseurl}}/{{page.version}}/reference/calicoctl/resources/ippool) for information on how to enable this for IP pools.
+
+2. Configure the VPC Routing Table
+
+   Alternatively, the VPC routing table can be programmed with container/pod IP
+   routes so that it can route container traffic natively.
+   This can be accomplished in two ways:
+
+   1. Deploy [canal][canal] and select the [aws-vpc backend][aws-vpc-backend],
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: Looks like you're missing the canal + aws-vpc-backend links.

File: master/reference/public-cloud/aws.md [hunk 32]
```diff
@@ -0,0 +1,113 @@
...
+
+To circumvent this issue, enable `ipip` and `nat-outgoing` on your IP pools ([as discussed above](#allow-traffic-across-different-subnets))
+
+#### Allow Workload-to-WAN Traffic
+
+In a private cloud where operators maintain full control of the networking
+fabric, NAT typically takes place at the edge router.
+
+AWS does not support NAT on outbound traffic from unknown sources
+(i.e. container/pod IPs).
```
Comment (caseydavenport) [inline master/reference/public-cloud/aws.md]: super picky nit, but this should be `e.g` instead of `i.e`

Might be easier to word as "AWS performs outbound NAT on traffic with the source address of a known EC2 instance, but does not support NAT on traffic from container/pod IP addresses.
